File: hireconnect-web/src/main/java/com/hireconnect/web/service/NotificationService.java
package com.hireconnect.web.service;

import java.util.List;

import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;

import com.hireconnect.web.dto.NotificationResponse;
import com.hireconnect.web.dto.PortalSession;
import com.hireconnect.web.support.GatewayClient;

@Service
public class NotificationService {

    private final GatewayClient gatewayClient;

    public NotificationService(GatewayClient gatewayClient) {
        this.gatewayClient = gatewayClient;
    }

    public List<NotificationResponse> getByUser(Integer userId, PortalSession session) {
        return gatewayClient.get("/api/v1/notifications/user/{userId}", session, new ParameterizedTypeReference<>() {
        }, userId);
    }

    public Integer getUnreadCount(Integer userId, PortalSession session) {
        return gatewayClient.get("/api/v1/notifications/user/{userId}/unread-count", session, Integer.class, userId);
    }

    public void markAsRead(Integer notificationId, PortalSession session) {
        gatewayClient.patch("/api/v1/notifications/{notificationId}/read", session, null, Void.class, notificationId);
    }

    public void markAllRead(Integer userId, PortalSession session) {
        gatewayClient.patch("/api/v1/notifications/user/{userId}/read-all", session, null, Void.class, userId);
    }
}

File: frontend/src/app/core/services/notification.service.ts
import { HttpClient, HttpParams, HttpResponse } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { API_ENDPOINTS } from '../constants/api.constants';
import { NotificationEvent, NotificationResponse } from '../models/notification.models';

@Injectable({ providedIn: 'root' })
export class NotificationService {
  private readonly http = inject(HttpClient);

  publishEvent(payload: NotificationEvent): Observable<void> {
    return this.http.post<void>(`${API_ENDPOINTS.notifications}/events`, payload);
  }

  getByUser(userId: number, isRead?: boolean | null): Observable<NotificationResponse[]> {
    let params = new HttpParams();
    if (isRead !== null && isRead !== undefined) {
      params = params.set('isRead', String(isRead));
    }
    return this.http.get<NotificationResponse[]>(`${API_ENDPOINTS.notifications}/user/${userId}`, { params });
  }

  markAsRead(notificationId: number): Observable<void> {
    return this.http.patch<void>(`${API_ENDPOINTS.notifications}/${notificationId}/read`, {});
  }

  markAllRead(userId: number): Observable<void> {
    return this.http.patch<void>(`${API_ENDPOINTS.notifications}/user/${userId}/read-all`, {});
  }

  deleteNotification(notificationId: number): Observable<void> {
    return this.http.delete<void>(`${API_ENDPOINTS.notifications}/${notificationId}`);
  }

  getUnreadCount(userId: number): Observable<number> {
    return this.http.get<number>(`${API_ENDPOINTS.notifications}/user/${userId}/unread-count`);
  }

  downloadOfferLetter(applicationId: number): Observable<HttpResponse<Blob>> {
    return this.http.get(`${API_ENDPOINTS.notifications}/offer-letters/${applicationId}`, {
      observe: 'response',
      responseType: 'blob',
    });
  }
}
